fuse_models/benchmark/benchmark_unicycle_2d_state_cost_function.cpp

```diff
@@ -0,0 +1,156 @@
+/*
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2019, Clearpath Robotics
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of the copyright holder nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <fuse_models/unicycle_2d_state_cost_function.h>
+#include <fuse_models/unicycle_2d_state_cost_functor.h>
+
+#include <benchmark/benchmark.h>
+
+#include <ceres/autodiff_cost_function.h>
+#include <Eigen/Dense>
+
+#include <vector>
+
+class Unicycle2DStateCostFunction : public benchmark::Fixture
+{
+public:
+  EIGEN_MAKE_ALIGNED_OPERATOR_NEW
+
+  Unicycle2DStateCostFunction()
+    : jacobians(num_parameter_blocks)
+    , J(num_parameter_blocks)
+  {
+    for (size_t i = 0; i < num_parameter_blocks; ++i)
+    {
+      J[i].resize(num_residuals, block_sizes[i]);
+      jacobians[i] = J[i].data();
+    }
+  }
+
+  // Analytic cost function
```

Comment:
I tried not to use global variables. For that reason I have a `benchmark::Fixture`. I also used a bunch of `static` attributes, so I'm not sure if this is cleaner/better at the end.

Reply:
In addition, since almost everything is `static`, I didn't have to implement any `SetUp` and `TearDown` methods.
But I wonder if I should maybe do some sort of cleanup, like `J.clear()` and `jacobians.clear()` to make sure things are working as expected. That would change the constructor, and it won't really test things are producing the correct results though. At the end I left it as is because I think the unit test for the analytic jacobians is already doing the job of testing it. This indeed shares some code with that unit test. I'm open to any suggestion to make things cleaner and more modular, if you think it's feasible and worth it.

Reply:
* I agree that the unit tests verify the correct values are computed, and the benchmark test measures the runtime. No need to mix the responsibilities.
* I didn't know about the Google Benchmark suite before now. This is great that you formalized the test.
* I'm not terribly concerned about sharing code between unit tests or between unit tests and benchmark tests. The code here is mostly just initialization anyway. No need to over complicate the structure.
* I'm not sure it really matters one way or the other about the static members. Assuming the benchmark suite works the same way as the gtest suite, the test fixture constructor should get called before the timing actually occurs, and each benchmark would use a clean test fixture object. I would expect you would get the exact same results with non-static member variables that were initialized in the constructor or setup() method. It is really just personal style preference at this point.